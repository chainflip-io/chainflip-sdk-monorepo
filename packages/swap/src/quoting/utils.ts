import assert from 'assert';
import BigNumber from 'bignumber.js';
import { filter, Observable, Subscription } from 'rxjs';
import { Side, type SwapInput } from '@/amm-addon';
import { InternalAsset, getAssetAndChain } from '@/shared/enums';
import { MarketMakerQuote, Leg as MarketMakerLeg } from './schemas';
import env from '../config/env';
import { getAssetPrice } from '../pricing';
import { assertUnreachable } from '../utils/function';
import { SwapRateArgs } from '../utils/statechain';

export type QuoteType = 'pool' | 'market_maker';

export const collectMakerQuotes = (
  requestId: string,
  expectedQuotes: number,
  quotes$: Observable<{ marketMaker: string; quote: MarketMakerQuote }>,
): Promise<MarketMakerQuote[]> => {
  if (expectedQuotes === 0) return Promise.resolve([]);

  const clientsReceivedQuotes = new Map<string, MarketMakerQuote>();

  return new Promise((resolve) => {
    let sub: Subscription;

    const complete = () => {
      sub.unsubscribe();
      resolve([...clientsReceivedQuotes.values()]);
    };

    sub = quotes$
      .pipe(filter(({ quote }) => quote.request_id === requestId))
      .subscribe(({ marketMaker, quote }) => {
        clientsReceivedQuotes.set(marketMaker, quote);
        if (clientsReceivedQuotes.size === expectedQuotes) complete();
      });

    setTimeout(complete, env.QUOTE_TIMEOUT);
  });
};

export const approximateIntermediateOutput = async (asset: InternalAsset, amount: string) => {
  const price = await getAssetPrice(asset);

  if (typeof price !== 'number') return null;

  return BigInt(new BigNumber(amount).times(price).toFixed(0));
};

type BaseAsset = Exclude<InternalAsset, 'Usdc'>;

export class Leg {
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint): Leg;
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint | null): Leg | null;
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint | null) {
    if (amount === null) return null;
    assert(baseAsset !== quoteAsset, 'baseAsset and quoteAsset must be different');
    assert(baseAsset === 'Usdc' || quoteAsset === 'Usdc', 'one of the assets must be USDC');
    return new Leg(baseAsset, quoteAsset, amount);
  }

  private constructor(
    private readonly baseAsset: InternalAsset,
    private readonly quoteAsset: InternalAsset,
    public amount: bigint,
  ) {}

  toPoolJSON(): SwapRateArgs {
    return {
      amount: this.amount,
      srcAsset: this.baseAsset,
      destAsset: this.quoteAsset,
    };
  }

  private getSide(): 'BUY' | 'SELL' {
    return this.quoteAsset !== 'Usdc' ? 'BUY' : 'SELL';
  }

  toMarketMakerJSON(): MarketMakerLeg {
    const side = this.getSide();
    let normalizedBaseAsset: BaseAsset;
    const normalizedQuoteAsset = 'Usdc';

    if (this.quoteAsset !== 'Usdc') {
      normalizedBaseAsset = this.quoteAsset;
    } else if (this.baseAsset !== 'Usdc') {
      normalizedBaseAsset = this.baseAsset;
    } else {
      return assertUnreachable('invalid leg');
    }

    return {
      base_asset: getAssetAndChain(normalizedBaseAsset),
      quote_asset: getAssetAndChain(normalizedQuoteAsset),
      amount: this.amount.toString(),
      side,
    };
  }

  toSwapInput(): SwapInput {
    return {
      side: this.getSide() === 'BUY' ? Side.Buy : Side.Sell,
      amount: this.amount,
      limitOrders: [],
    };
  }
}
