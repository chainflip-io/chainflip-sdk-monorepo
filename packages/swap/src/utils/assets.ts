import { SupportedAsset, supportedAsset } from '@/shared/enums';

export const isSupportedAsset = (value: string): value is SupportedAsset =>
  supportedAsset.safeParse(value).success;

export function assertSupportedAsset(
  value: string,
): asserts value is SupportedAsset {
  if (!isSupportedAsset(value)) {
    throw new Error(`received invalid asset "${value}"`);
  }
}

export const decimalPlaces: Record<SupportedAsset, number> = {
  DOT: 10,
  ETH: 18,
  FLIP: 18,
  USDC: 6,
  BTC: 8,
};
