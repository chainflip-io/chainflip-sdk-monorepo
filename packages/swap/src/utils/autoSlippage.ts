import BigNumber from 'bignumber.js';
import { assetConstants, chainConstants } from '@/shared/enums';
import { getDeployedLiquidity, getUndeployedLiquidity } from './pools';
import { getRequiredBlockConfirmations } from './rpc';
import { InternalAsset } from '../client';

const getLiquidityAutoSlippage = async (
  srcAsset: InternalAsset,
  destAsset: InternalAsset,
  amount: bigint,
) => {
  const deployedLiquidity = await getDeployedLiquidity(srcAsset, destAsset);
  const liquidityRatio = new BigNumber(amount.toString()).div(deployedLiquidity.toString());

  if (liquidityRatio.isGreaterThan(0.2)) {
    return liquidityRatio.multipliedBy(2);
  }
  return new BigNumber(0);
};

const getDepositTimeAutoSlippage = async (srcAsset: InternalAsset, isBoosted: boolean) => {
  const { chain } = assetConstants[srcAsset];
  const { blockTimeSeconds } = chainConstants[chain];
  const blockConfirmations = isBoosted ? 1 : (await getRequiredBlockConfirmations(srcAsset)) ?? 0;
  const depositTimeMinutes = (blockConfirmations * blockTimeSeconds) / 60;

  if (['Solana', 'Arbitrum'].includes(chain)) {
    return depositTimeMinutes * 0.01;
  }
  if (['Ethereum', 'Polkadot'].includes(chain)) {
    return depositTimeMinutes * 0.05;
  }
  return depositTimeMinutes * 0.1;
};

const getUndeployedLiquidityAutoSlippage = async (asset: InternalAsset, amount: bigint) => {
  const undeployedLiquidity = await getUndeployedLiquidity(asset);
  if (undeployedLiquidity >= amount) {
    return new BigNumber(-0.5);
  }

  const undeployedRatio = new BigNumber(undeployedLiquidity.toString()).div(amount.toString());
  if (undeployedRatio.gt(0.5)) {
    return new BigNumber(-undeployedRatio * 0.1);
  }

  return new BigNumber(0);
};

export const calculateRecommendedSlippage = async ({
  srcAsset,
  destAsset,
  intermediateAmount,
  egressAmount,
  boostFeeBps,
  dcaChunks,
}: {
  srcAsset: InternalAsset;
  destAsset: InternalAsset;
  intermediateAmount?: bigint;
  egressAmount: bigint;
  boostFeeBps?: number;
  dcaChunks: number;
}) => {
  const minSlippage = 0.1;
  const baseSlippage = 1;

  const assetTo = srcAsset === 'Usdc' || destAsset === 'Usdc' ? destAsset : 'Usdc';

  const [deployedLiquiditySlippage, timeSlippage, undeployedLiquiditySlippage] = await Promise.all([
    await getLiquidityAutoSlippage(
      srcAsset,
      assetTo,
      (intermediateAmount ?? egressAmount) * BigInt(dcaChunks),
    ),
    await getDepositTimeAutoSlippage(srcAsset, Boolean(boostFeeBps && boostFeeBps > 0)),
    await getUndeployedLiquidityAutoSlippage(
      intermediateAmount ? 'Usdc' : destAsset,
      (intermediateAmount ?? egressAmount) * BigInt(dcaChunks),
    ),
  ]);

  const calculatedSlippage = deployedLiquiditySlippage
    .plus(timeSlippage)
    .plus(undeployedLiquiditySlippage)
    .plus(baseSlippage);

  return Math.max(minSlippage, Math.round(calculatedSlippage.toNumber() * 100) / 100);
};
