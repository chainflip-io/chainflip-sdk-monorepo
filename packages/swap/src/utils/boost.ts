import { ONE_IN_PIP, bigintMin, getPipAmountFromAmount } from '@/shared/functions';
import { BoostPoolsDepth } from '@/shared/rpc';

export const getBoostFeeBpsForAmount = async ({
  amount,
  assetBoostPoolsDepth,
}: {
  amount: bigint;
  assetBoostPoolsDepth: BoostPoolsDepth;
}): Promise<number | undefined> => {
  assetBoostPoolsDepth.sort((a, b) => (a.tier < b.tier ? -1 : 1));

  let _amount = amount;
  let feeAmount = 0n;

  for (const poolDepth of assetBoostPoolsDepth) {
    const poolAvailableAmount = poolDepth.availableAmount;

    const amountToBeUsedFromPool = bigintMin(_amount, poolAvailableAmount);
    feeAmount += getPipAmountFromAmount(amountToBeUsedFromPool, poolDepth.tier);
    _amount -= amountToBeUsedFromPool;
  }

  // Not enough liquidity in the boost pools
  if (_amount > 0) return undefined;

  return Number((feeAmount * BigInt(ONE_IN_PIP)) / BigInt(amount));
};
