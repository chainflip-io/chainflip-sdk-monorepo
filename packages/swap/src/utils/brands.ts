/**
 * Branded types are used to create types that are compatible with a base type,
 * but are not assignable to the base type. This is useful for creating types
 * that are more specific than the base type and should not be used interchangeably.
 *
 * @see https://www.learningtypescript.com/articles/branded-types
 */
export type Brand<T, B> = T & { __brand: B };
