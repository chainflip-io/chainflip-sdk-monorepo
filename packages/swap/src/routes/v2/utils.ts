import { getAssetAndChain } from '@/shared/enums';
import { assertUnreachable } from '@/shared/functions';
import { StateV2 } from './swap';
import prisma, {
  Prisma,
  Swap,
  Egress,
  Broadcast,
  SwapFee,
  IgnoredEgress,
  FailedSwap,
} from '../../client';
import { getPendingBroadcast } from '../../ingress-egress-tracking';
import { failedSwapMessage } from '../../utils/swap';
import { FailureMode } from '../swap';

export const depositChannelInclude = {
  failedBoosts: true,
  failedSwaps: true,
  affiliates: {
    select: {
      account: true,
      commissionBps: true,
    },
  },
} as const;

export const swapRequestInclude = {
  swaps: { include: { fees: true }, orderBy: { nativeId: 'desc' } },
  egress: { include: { broadcast: true } },
  refundEgress: { include: { broadcast: true } },
  fees: true,
  ignoredEgresses: true,
  swapDepositChannel: {
    include: depositChannelInclude,
  },
} as const;

export const getSwapFields = (swap: Swap & { fees: SwapFee[] }) => ({
  swapInputAmount: swap.swapInputAmount.toString(),
  swapOutputAmount: swap.swapOutputAmount?.toString(),
  scheduledAt: swap.swapScheduledAt.valueOf(),
  scheduledBlockIndex: swap.swapScheduledBlockIndex ?? undefined,
  executedAt: swap.swapExecutedAt?.valueOf(),
  executedBlockIndex: swap.swapExecutedBlockIndex ?? undefined,
  retryCount: swap.retryCount,
  latestSwapRescheduledAt: swap.latestSwapRescheduledAt?.valueOf(),
  latestSwapRescheduledBlockIndex: swap.latestSwapRescheduledBlockIndex ?? undefined,
  fees: swap.fees.map((fee) => ({
    type: fee.type,
    ...getAssetAndChain(fee.asset),
    amount: fee.amount.toString(),
  })),
});

export const getDepositIgnoredFailedState = (failedSwap: FailedSwap) => ({
  failedAt: failedSwap.failedAt.valueOf(),
  failedAtBlockIndex: failedSwap.failedBlockIndex,
  mode: FailureMode.IngressIgnored,
  reason: {
    name: failedSwap.reason,
    message: failedSwapMessage[failedSwap.reason],
  },
});

export const getEgressFailureState = async (
  ignoredEgress: IgnoredEgress | undefined,
  broadcast: Broadcast | null | undefined,
  type: IgnoredEgress['type'] | undefined,
) => {
  const abortedBroadcast = broadcast?.abortedAt?.valueOf();
  const failedAt = abortedBroadcast ?? ignoredEgress?.ignoredAt?.valueOf();
  const failedAtBlockIndex = broadcast?.abortedBlockIndex ?? ignoredEgress?.ignoredBlockIndex;

  let error;
  let failureMode;

  if (failedAt && failedAtBlockIndex) {
    error = {
      name: 'Unknown',
      message: 'An unknown error occurred',
    };
    if (ignoredEgress) {
      switch (ignoredEgress.type) {
        case 'REFUND':
          failureMode = FailureMode.RefundEgressIgnored;
          break;
        case 'SWAP':
          failureMode = FailureMode.EgressIgnored;
          break;
        default:
          assertUnreachable(ignoredEgress.type);
      }
      const [stateChainError] = await Promise.all([
        prisma.stateChainError.findUniqueOrThrow({
          where: { id: ignoredEgress.stateChainErrorId },
        }),
      ]);
      error = {
        name: stateChainError.name,
        message: stateChainError.docs,
      };
    } else if (abortedBroadcast) {
      const message =
        type === 'SWAP' ? 'The swap broadcast was aborted' : 'The refund broadcast was aborted';
      failureMode = FailureMode.BroadcastAborted;
      error = {
        name: 'BroadcastAborted',
        message,
      };
    }
    return {
      failedAt,
      failedAtBlockIndex,
      mode: failureMode,
      reason: error,
    };
  }
  return undefined;
};

export const getEgressStatusFields = async (
  egress: Egress | null | undefined,
  broadcast: Broadcast | null | undefined,
  ignoredEgresses: IgnoredEgress[] | undefined,
  type: IgnoredEgress['type'] | undefined,
  egressTrackerTxRef: string | null | undefined,
) => {
  const ignoredEgress = ignoredEgresses?.find((e) => e.type === type);
  const failureState = await getEgressFailureState(ignoredEgress, broadcast, type);
  return {
    ...(egress && {
      outputAmount: egress.amount?.toFixed(),
      scheduledAt: egress.scheduledAt?.valueOf(),
      scheduledBlockIndex: egress.scheduledBlockIndex ?? undefined,
    }),
    ...(broadcast && {
      sentAt: broadcast?.succeededAt?.valueOf(),
      sentAtBlockIndex: broadcast?.succeededBlockIndex ?? undefined,
      sentTxRef: broadcast?.transactionRef ?? egressTrackerTxRef,
    }),
    ...(failureState && { failure: failureState }),
    ...(ignoredEgress && { ignoredAmount: ignoredEgress.amount?.toFixed() }),
  };
};

export const getSwapState = async (
  failedSwap: FailedSwap | null | undefined,
  ignoredEgresses: IgnoredEgress[] | undefined,
  swapRequest:
    | Prisma.SwapRequestGetPayload<{
        include: typeof swapRequestInclude;
      }>
    | undefined
    | null,
) => {
  let state: StateV2 | undefined;
  let swapEgressTrackerTxRef: string | null | undefined;
  let refundEgressTrackerTxRef: string | null | undefined;
  const swapEgress = swapRequest?.egress;
  const refundEgress = swapRequest?.refundEgress;
  const egress = swapEgress ?? refundEgress;

  if (
    failedSwap ||
    (ignoredEgresses && ignoredEgresses.length > 0) ||
    egress?.broadcast?.abortedAt
  ) {
    state = StateV2.Failed;
  } else if (egress?.broadcast?.succeededAt) {
    state = StateV2.Completed;
  } else if (egress?.broadcast) {
    if (swapEgress?.broadcast) {
      state = StateV2.Sending;
      const pendingSwapBroadcast = await getPendingBroadcast(swapEgress.broadcast);
      if (pendingSwapBroadcast) {
        state = StateV2.Sent;
        swapEgressTrackerTxRef = pendingSwapBroadcast.tx_ref;
      }
    }
    if (refundEgress?.broadcast) {
      const pendingRefundBroadcast = await getPendingBroadcast(refundEgress.broadcast);
      if (pendingRefundBroadcast) {
        state = StateV2.Sent;
        refundEgressTrackerTxRef = pendingRefundBroadcast.tx_ref;
      }
    }
  } else if (egress) {
    state = StateV2.Sending;
  } else if (swapRequest?.swaps.some((s) => s.swapScheduledAt)) {
    state = StateV2.Swapping;
  } else {
    state = StateV2.Receiving;
  }

  return {
    state,
    swapEgressTrackerTxRef,
    refundEgressTrackerTxRef,
  };
};
