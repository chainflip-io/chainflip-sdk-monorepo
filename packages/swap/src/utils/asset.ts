import { ChainflipAsset } from '@chainflip/utils/chainflip';

export const isAssetHub = (asset: ChainflipAsset): asset is 'HubDot' | 'HubUsdc' | 'HubUsdt' =>
  asset === 'HubDot' || asset === 'HubUsdc' || asset === 'HubUsdt';
