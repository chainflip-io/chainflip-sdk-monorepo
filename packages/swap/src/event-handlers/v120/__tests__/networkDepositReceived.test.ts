import { networkDepositReceivedBtcMockV120 as networkDepositReceivedBtcMock } from '@/swap/event-handlers/__tests__/utils';
import prisma from '../../../client';
import {
  depositReceivedArgs as depositReceivedArgsV120,
  networkDepositReceived as networkDepositReceivedV120,
} from '../networkDepositReceived';

describe('depositReceivedV120', () => {
  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "private"."DepositChannel" CASCADE`;
    await prisma.$queryRaw`TRUNCATE TABLE "SwapDepositChannel", "Swap" CASCADE`;
  });

  it('should update the values for an existing swap', async () => {
    await prisma.depositChannel.create({
      data: {
        srcChain: 'Bitcoin',
        depositAddress:
          'tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6',
        channelId: 3,
        issuedBlock: 0,
        isSwapping: true,
      },
    });

    const swapDepositChannel = await prisma.swapDepositChannel.create({
      data: {
        srcAsset: 'BTC',
        srcChain: 'Bitcoin',
        srcChainExpiryBlock: 100,
        depositAddress:
          'tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6',
        expectedDepositAmount: 0,
        destAddress: '0x6fd76a7699e6269af49e9c63f01f61464ab21d1c',
        brokerCommissionBps: 0,
        destAsset: 'ETH',
        channelId: 3,
        issuedBlock: 0,
        swaps: {
          create: {
            swapInputAmount: '100000',
            depositAmount: '100000',
            srcAsset: 'BTC',
            destAsset: 'ETH',
            destAddress: '0x6fd76a7699e6269af49e9c63f01f61464ab21d1c',
            type: 'SWAP',
            nativeId: 1,
            depositReceivedAt: new Date(1670337099000),
            depositReceivedBlockIndex: `0-15`,
          } as any,
        },
      },
    });

    await prisma.$transaction(async (txClient) => {
      await networkDepositReceivedV120({
        prisma: txClient,
        block: networkDepositReceivedBtcMock().block as any,
        event: networkDepositReceivedBtcMock().eventContext.event as any,
      });
    });

    const swap = await prisma.swap.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
      include: { fees: true },
    });

    expect(swap.depositAmount.toString()).toBe('110000');
    expect(swap).toMatchSnapshot({
      id: expect.any(BigInt),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      swapDepositChannelId: expect.any(BigInt),
      fees: [{ id: expect.any(BigInt), swapId: expect.any(BigInt) }],
    });
  });

  it('should update the values for an existing ccm principal swap', async () => {
    await prisma.depositChannel.create({
      data: {
        srcChain: 'Bitcoin',
        depositAddress:
          'tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6',
        channelId: 3,
        issuedBlock: 0,
        isSwapping: true,
      },
    });

    const swapDepositChannel = await prisma.swapDepositChannel.create({
      data: {
        srcAsset: 'BTC',
        srcChain: 'Bitcoin',
        srcChainExpiryBlock: 100,
        depositAddress:
          'tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6',
        expectedDepositAmount: 0,
        destAddress: '0x6fd76a7699e6269af49e9c63f01f61464ab21d1c',
        brokerCommissionBps: 0,
        destAsset: 'ETH',
        channelId: 3,
        issuedBlock: 0,
        swaps: {
          create: {
            swapInputAmount: '100000',
            depositAmount: '100000',
            srcAsset: 'BTC',
            destAsset: 'ETH',
            destAddress: '0x6fd76a7699e6269af49e9c63f01f61464ab21d1c',
            type: 'PRINCIPAL',
            nativeId: 1,
            depositReceivedAt: new Date(1670337099000),
            depositReceivedBlockIndex: `0-15`,
          } as any,
        },
      },
    });

    await prisma.$transaction(async (txClient) => {
      await networkDepositReceivedV120({
        prisma: txClient,
        block: networkDepositReceivedBtcMock().block as any,
        event: networkDepositReceivedBtcMock({
          __kind: 'CcmTransfer',
          principalSwapId: '1',
        }).eventContext.event as any,
      });
    });

    const swap = await prisma.swap.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
      include: { fees: true },
    });

    expect(swap.depositAmount.toString()).toBe('110000');
    expect(swap).toMatchSnapshot({
      id: expect.any(BigInt),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      swapDepositChannelId: expect.any(BigInt),
      fees: [{ id: expect.any(BigInt), swapId: expect.any(BigInt) }],
    });
  });

  it('should update the values for an existing ccm gas swap', async () => {
    await prisma.depositChannel.create({
      data: {
        srcChain: 'Bitcoin',
        depositAddress:
          'tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6',
        channelId: 3,
        issuedBlock: 0,
        isSwapping: true,
      },
    });

    const swapDepositChannel = await prisma.swapDepositChannel.create({
      data: {
        srcAsset: 'BTC',
        srcChain: 'Bitcoin',
        srcChainExpiryBlock: 100,
        depositAddress:
          'tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6',
        expectedDepositAmount: 0,
        destAddress: '0x6fd76a7699e6269af49e9c63f01f61464ab21d1c',
        brokerCommissionBps: 0,
        destAsset: 'ETH',
        channelId: 3,
        issuedBlock: 0,
        swaps: {
          create: {
            swapInputAmount: '100000',
            depositAmount: '100000',
            srcAsset: 'BTC',
            destAsset: 'ETH',
            destAddress: '0x6fd76a7699e6269af49e9c63f01f61464ab21d1c',
            type: 'PRINCIPAL',
            nativeId: 1,
            depositReceivedAt: new Date(1670337099000),
            depositReceivedBlockIndex: `0-15`,
          } as any,
        },
      },
    });

    await prisma.$transaction(async (txClient) => {
      await networkDepositReceivedV120({
        prisma: txClient,
        block: networkDepositReceivedBtcMock().block as any,
        event: networkDepositReceivedBtcMock({
          __kind: 'CcmTransfer',
          gasSwapId: '1',
        }).eventContext.event as any,
      });
    });

    const swap = await prisma.swap.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
      include: { fees: true },
    });

    expect(swap.depositAmount.toString()).toBe('110000');
    expect(swap).toMatchSnapshot({
      id: expect.any(BigInt),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      swapDepositChannelId: expect.any(BigInt),
      fees: [{ id: expect.any(BigInt), swapId: expect.any(BigInt) }],
    });
  });

  it('should throw if there is no swap for deposit channel', async () => {
    await prisma.depositChannel.create({
      data: {
        srcChain: 'Bitcoin',
        depositAddress:
          'tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6',
        channelId: 3,
        issuedBlock: 0,
        isSwapping: true,
      },
    });

    await prisma.swapDepositChannel.create({
      data: {
        srcAsset: 'BTC',
        srcChain: 'Bitcoin',
        srcChainExpiryBlock: 100,
        depositAddress:
          'tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6',
        expectedDepositAmount: 0,
        destAddress: '0x6fd76a7699e6269af49e9c63f01f61464ab21d1c',
        brokerCommissionBps: 0,
        destAsset: 'ETH',
        channelId: 3,
        issuedBlock: 0,
      },
    });

    await expect(
      prisma.$transaction(async (txClient) => {
        await networkDepositReceivedV120({
          prisma: txClient,
          block: networkDepositReceivedBtcMock().block as any,
          event: networkDepositReceivedBtcMock().eventContext.event as any,
        });
      }),
    ).rejects.toThrow();
  });

  it('should throw if the swapId does not match an existing swap', async () => {
    await prisma.depositChannel.create({
      data: {
        srcChain: 'Bitcoin',
        depositAddress:
          'tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6',
        channelId: 3,
        issuedBlock: 0,
        isSwapping: true,
      },
    });

    await prisma.swapDepositChannel.create({
      data: {
        srcAsset: 'BTC',
        srcChain: 'Bitcoin',
        srcChainExpiryBlock: 100,
        depositAddress:
          'tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6',
        expectedDepositAmount: 0,
        destAddress: '0x6fd76a7699e6269af49e9c63f01f61464ab21d1c',
        brokerCommissionBps: 0,
        destAsset: 'ETH',
        channelId: 3,
        issuedBlock: 0,
        swaps: {
          create: {
            swapInputAmount: '100000',
            depositAmount: '100000',
            srcAsset: 'BTC',
            destAsset: 'ETH',
            destAddress: '0x6fd76a7699e6269af49e9c63f01f61464ab21d1c',
            type: 'SWAP',
            nativeId: 100,
            depositReceivedAt: new Date(1670337099000),
            depositReceivedBlockIndex: `0-15`,
          } as any,
        },
      },
    });

    await expect(
      prisma.$transaction(async (txClient) => {
        await networkDepositReceivedV120({
          prisma: txClient,
          block: networkDepositReceivedBtcMock().block as any,
          event: networkDepositReceivedBtcMock().eventContext.event as any,
        });
      }),
    ).rejects.toThrow();
  });
});

describe('depositReceivedArgsV120V120', () => {
  it.each([
    {
      asset: { __kind: 'Btc' },
      amount: '1000000',
      depositAddress: {
        value:
          '0x69e988bde97e4b988f1d11fa362118c4bdf5e32c45a6e7e89fde3106b764bebd',
        __kind: 'Taproot',
      },
      depositDetails: {
        txId: '0x14fd88f956c399e64356546fea41ba234670a7b63c8e2b7e81c8f1ae9011b0d7',
        vout: 1,
      },
      action: {
        __kind: 'Swap',
        swapId: '1',
      },
      ingressFee: '1000',
    },
    {
      asset: { __kind: 'Flip' },
      amount: '9853636405123772134',
      depositAddress: '0xe0c0ca3540ddd2fc6244e62aa8c8f70c7021ff7f',
      action: {
        __kind: 'Swap',
        swapId: '1',
      },
      ingressFee: '1000',
    },
    {
      asset: { __kind: 'Usdc' },
      amount: '1000000000',
      depositAddress: '0x9a53bd378c459f71a74acd96bfcd64ed96d92a8e',
      action: {
        __kind: 'Swap',
        swapId: '1',
      },
      ingressFee: '1000',
    },
    {
      asset: { __kind: 'Eth' },
      amount: '100000000000000000',
      depositAddress: '0xf7b277413fd3e1f1d1e631b1b121443889e46719',
      action: {
        __kind: 'Swap',
        swapId: '1',
      },
      ingressFee: '1000',
    },
    {
      asset: { __kind: 'Dot' },
      amount: '30000000000',
      depositAddress:
        '0x2d369b6db7f9dc6f332ca5887208d5814dcd759a516ee2507f9582d8b25d7b97',
      action: {
        __kind: 'Swap',
        swapId: '1',
      },
      ingressFee: '1000',
    },
    {
      asset: { __kind: 'Eth' },
      amount: '100000000000000000',
      depositAddress: '0xf7b277413fd3e1f1d1e631b1b121443889e46719',
      action: {
        __kind: 'CcmTransfer',
        principalSwapId: '1',
      },
      ingressFee: '1000',
    },
    {
      asset: { __kind: 'Eth' },
      amount: '100000000000000000',
      depositAddress: '0xf7b277413fd3e1f1d1e631b1b121443889e46719',
      action: {
        __kind: 'CcmTransfer',
        gasSwapId: '1',
      },
      ingressFee: '1000',
    },
    {
      asset: { __kind: 'Eth' },
      amount: '100000000000000000',
      depositAddress: '0xf7b277413fd3e1f1d1e631b1b121443889e46719',
      action: {
        __kind: 'NoAction',
      },
      ingressFee: '1000',
    },
  ])('parses the event', (args) => {
    expect(depositReceivedArgsV120.safeParse(args).success).toBe(true);
  });
});
