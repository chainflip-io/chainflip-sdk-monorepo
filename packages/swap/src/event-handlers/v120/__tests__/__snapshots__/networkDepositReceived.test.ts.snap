// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`depositReceivedV120 should update the values for an existing ccm gas swap 1`] = `
{
  "ccmDepositReceivedBlockIndex": null,
  "ccmGasBudget": null,
  "ccmMessage": null,
  "createdAt": Any<Date>,
  "depositAmount": "110000",
  "depositBoostedAt": null,
  "depositBoostedBlockIndex": null,
  "depositReceivedAt": 2022-12-06T14:31:39.000Z,
  "depositReceivedBlockIndex": "0-15",
  "depositTransactionRef": null,
  "destAddress": "0x6fd76a7699e6269af49e9c63f01f61464ab21d1c",
  "destAsset": "Eth",
  "effectiveBoostFeeBps": null,
  "egressId": null,
  "fees": [
    {
      "amount": "1000",
      "asset": "Btc",
      "id": Any<BigInt>,
      "swapId": Any<BigInt>,
      "type": "INGRESS",
    },
  ],
  "id": Any<BigInt>,
  "intermediateAmount": null,
  "latestSwapScheduledAt": 2022-12-06T14:31:39.000Z,
  "latestSwapScheduledBlockIndex": "0-15",
  "nativeId": 1n,
  "prewitnessedDepositId": null,
  "retryCount": 0,
  "srcAsset": "Btc",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": null,
  "swapExecutedBlockIndex": null,
  "swapInputAmount": "100000",
  "swapOutputAmount": null,
  "type": "PRINCIPAL",
  "updatedAt": Any<Date>,
}
`;

exports[`depositReceivedV120 should update the values for an existing ccm principal swap 1`] = `
{
  "ccmDepositReceivedBlockIndex": null,
  "ccmGasBudget": null,
  "ccmMessage": null,
  "createdAt": Any<Date>,
  "depositAmount": "110000",
  "depositBoostedAt": null,
  "depositBoostedBlockIndex": null,
  "depositReceivedAt": 2022-12-06T14:31:39.000Z,
  "depositReceivedBlockIndex": "0-15",
  "depositTransactionRef": null,
  "destAddress": "0x6fd76a7699e6269af49e9c63f01f61464ab21d1c",
  "destAsset": "Eth",
  "effectiveBoostFeeBps": null,
  "egressId": null,
  "fees": [
    {
      "amount": "1000",
      "asset": "Btc",
      "id": Any<BigInt>,
      "swapId": Any<BigInt>,
      "type": "INGRESS",
    },
  ],
  "id": Any<BigInt>,
  "intermediateAmount": null,
  "latestSwapScheduledAt": 2022-12-06T14:31:39.000Z,
  "latestSwapScheduledBlockIndex": "0-15",
  "nativeId": 1n,
  "prewitnessedDepositId": null,
  "retryCount": 0,
  "srcAsset": "Btc",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": null,
  "swapExecutedBlockIndex": null,
  "swapInputAmount": "100000",
  "swapOutputAmount": null,
  "type": "PRINCIPAL",
  "updatedAt": Any<Date>,
}
`;

exports[`depositReceivedV120 should update the values for an existing swap 1`] = `
{
  "ccmDepositReceivedBlockIndex": null,
  "ccmGasBudget": null,
  "ccmMessage": null,
  "createdAt": Any<Date>,
  "depositAmount": "110000",
  "depositBoostedAt": null,
  "depositBoostedBlockIndex": null,
  "depositReceivedAt": 2022-12-06T14:31:39.000Z,
  "depositReceivedBlockIndex": "0-15",
  "depositTransactionRef": null,
  "destAddress": "0x6fd76a7699e6269af49e9c63f01f61464ab21d1c",
  "destAsset": "Eth",
  "effectiveBoostFeeBps": null,
  "egressId": null,
  "fees": [
    {
      "amount": "1000",
      "asset": "Btc",
      "id": Any<BigInt>,
      "swapId": Any<BigInt>,
      "type": "INGRESS",
    },
  ],
  "id": Any<BigInt>,
  "intermediateAmount": null,
  "latestSwapScheduledAt": 2022-12-06T14:31:39.000Z,
  "latestSwapScheduledBlockIndex": "0-15",
  "nativeId": 1n,
  "prewitnessedDepositId": null,
  "retryCount": 0,
  "srcAsset": "Btc",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": null,
  "swapExecutedBlockIndex": null,
  "swapInputAmount": "100000",
  "swapOutputAmount": null,
  "type": "SWAP",
  "updatedAt": Any<Date>,
}
`;
